# Cleaning Used Car Dataset from eBay Kleinanzeigen

In this notebook, we will clean and analyse a dataset of used cars from the classifieds section of the German eBay website. A redacted version including 50,000 datapoints (down from the 370,000 datapoints per the [original](https://data.world/data-society/used-cars-data)) will be used in this project.

We will begin by importing the pandas and Numpy libraries, and loading the dataset into a dataframe.

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding = 'latin1') # three most common encodings utf8, latin1, and windows1252 were tested.

In [2]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [3]:
autos.info() # provides an overview of the dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

The dataframe contains 20 columns, 5 of which are classified as integers and the remaining 20 as object, likely strings.

There also appear to be null objects in columns vehicleType, gearbox, model, fuelType, and brand.

The column names are in camelcase instead of snakecase.

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We will manually change the column labels to introduce snakecasing and underscores for improved readability.

In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'no_pictures', 'postal_code',
       'last_seen'] 
#.map method may be used as well for clarity.

In [6]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  no_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

We will carry out some basic data exploration to identify the cleaning tasks that need to be completed prior to analysis:
1. Confirm if there's any text column with identical or almost identical data - the information will be dropped and result kept to keep the dataframe lean.
2. Check if there's any numeric data that need to be cleaned and converted to int/float type.

In [7]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-08 10:40:35  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerPS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  no_pictures   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

Based on the *describe* results:
- *seller* and *offer_type* columns may be checked and dropped as each column only contains two unique characters, with the dominant ones being 'privat' and 'Angebot' respectively.
- The most common reuslts in *price* column is \\$0, which cannot logically be the actual sales value. This column should be investigated furhter to identify if the \\$0 should be substituted with 'auction', 'custom offer', or other pricing arrangements.
- The data in the *odometer* column ay be processed and stored as numeric values. The column label should be mofified to reflect the units.

In [8]:
autos['odometer'].unique() # all the unique entries in odometer columns have ',' and 'km'.

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [9]:
odo_km = autos['odometer'].str.replace('km','').str.replace(',','').astype(int)

In [10]:
autos['odometer'] = odo_km

In [11]:
autos = autos.rename(columns = {'odometer':'odometer_km'}) # this renaming method is an alternative to direct assignment.

In [12]:
price_val = autos['price'].str.replace('$','').str.replace(',','').astype(int)

In [13]:
autos['price'] = price_val # changing the price column to numeric values.

With the price column now converted into int64 format, we can now analyse it further.

In [14]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [15]:
price_count = autos['price'].value_counts() # perform value counts on the price column to identify frequency of the unique entries.

In [16]:
price_count.sort_index(ascending = False).head(10) # sorting the price value counts by index magnitude to check for unrealistic prices.

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64

In [17]:
price_count.sort_index(ascending = True).head(10)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price, dtype: int64

Aside from the unrealistic $0 entries we have identified earlier, there are also some items listed at extremely high values (e.g. 99999999, 27322222). It is more likely that these high-profile sales would take place through registered dealers instead of a marketplace like eBay.

We will set a minimum threshold of \\$1,100 (25% checkpoint of the original price entries) and a maximum threshold of \\$200,000 (equivalent to approximately four times the average annual wage in Germany, in USD, [Statistia](https://www.statista.com/statistics/416207/average-annual-wages-germany-y-on-y-in-euros/#:~:text=Average%20annual%20wages%20in%20Germany%202000%2D2019&text=As%20of%202019%2C%20the%20average,Euros%20when%20compared%20with%202000.)).

There may be situations where normal datapoints are excluded by this range. For instance, buyers may be eager to get rid of their cars and set the prices extremely low. For the purpose of this project, however, the boundaries set based on the price distribution and national average wage would be sufficient.

In [18]:
price_filtered_index = autos['price'].between(1100,200000)

In [19]:
price_filtered = autos['price'][price_filtered_index]

There are 37,816 entries in the price filtered series - we have retained approximately 75% of the entries after filtering.

In [20]:
price_filtered.describe() # the values appear to be in line with the boundary values we have set.

count     37816.000000
mean       7328.917204
std        8859.541797
min        1100.000000
25%        2300.000000
50%        4500.000000
75%        9000.000000
max      198000.000000
Name: price, dtype: float64

In [21]:
autos_filtered = autos[price_filtered_index]

We will now examine the *odometer_km* column.

In [22]:
autos_filtered['odometer_km'].describe()

count     37816.000000
mean     122434.154855
std       40911.709533
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [23]:
autos_filtered['odometer_km'].value_counts()

150000    22639
125000     4285
100000     1846
90000      1555
80000      1326
70000      1153
60000      1093
50000       981
40000       795
30000       744
20000       685
5000        487
10000       227
Name: odometer_km, dtype: int64

The *odometer_km* appears to provide the odometer readings for the vehicles, classified into ceiling ranges (i.e. 0~10,000 km will fall into 10,000 km). There doesn't appear to be any invalid entries - the majority of the cars being sold on eBay marketplace would be used cars. Both the lower limit of 10,000km and 150,000km are reasonable readings on car odometers.

We will move onto cleaning the date columns and converting them into numeric values that are more suitable for analysis.

Begin by printing out the first five rows of the *date_crawled*, *ad_created*, and *last_seen* columns. This gives an understanding of the datetime format used.

In [24]:
autos_filtered[['date_crawled','ad_created','last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We will extract only the date values (first 10 string characters) for each column. The distribution of dates can then be analysed individually.

In [25]:
#percentage of sales listed on each date for date_crawled, sorted from earliest to latest.
autos_filtered['date_crawled'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.025703
2016-03-06    0.013962
2016-03-07    0.035461
2016-03-08    0.032314
2016-03-09    0.032579
2016-03-10    0.033372
2016-03-11    0.033028
2016-03-12    0.037524
2016-03-13    0.016078
2016-03-14    0.036783
2016-03-15    0.033451
2016-03-16    0.029115
2016-03-17    0.030146
2016-03-18    0.012931
2016-03-19    0.035144
2016-03-20    0.038158
2016-03-21    0.037365
2016-03-22    0.032685
2016-03-23    0.032103
2016-03-24    0.028744
2016-03-25    0.030331
2016-03-26    0.033108
2016-03-27    0.031442
2016-03-28    0.035276
2016-03-29    0.034139
2016-03-30    0.032896
2016-03-31    0.031415
2016-04-01    0.034536
2016-04-02    0.036334
2016-04-03    0.039269
2016-04-04    0.036704
2016-04-05    0.013195
2016-04-06    0.003200
2016-04-07    0.001507
Name: date_crawled, dtype: float64

The distribution of data appears to be relatively even. There is no N/A entry. Data cleaning is unnecessary.

In [26]:
#percentage of sales listed on each date for ad_created, sorted from earliest to latest.
autos_filtered['ad_created'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2015-06-11    0.000026
2015-08-10    0.000026
2015-09-09    0.000026
2015-11-10    0.000026
2015-12-30    0.000026
                ...   
2016-04-03    0.039560
2016-04-04    0.037154
2016-04-05    0.011794
2016-04-06    0.003305
2016-04-07    0.001322
Name: ad_created, Length: 73, dtype: float64

More ads appear to have been created recently. There is no N/A entry. Data cleaning is unnecessary.

In [27]:
#percentage of sales listed on each date for last_seen, sorted from earliest to latest.
autos_filtered['last_seen'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.001084
2016-03-06    0.003570
2016-03-07    0.004469
2016-03-08    0.006294
2016-03-09    0.008832
2016-03-10    0.009678
2016-03-11    0.011635
2016-03-12    0.022239
2016-03-13    0.008356
2016-03-14    0.012006
2016-03-15    0.014835
2016-03-16    0.015496
2016-03-17    0.026153
2016-03-18    0.007351
2016-03-19    0.014518
2016-03-20    0.019674
2016-03-21    0.019383
2016-03-22    0.020838
2016-03-23    0.017982
2016-03-24    0.018511
2016-03-25    0.017744
2016-03-26    0.015946
2016-03-27    0.014015
2016-03-28    0.019383
2016-03-29    0.020573
2016-03-30    0.023403
2016-03-31    0.022795
2016-04-01    0.023165
2016-04-02    0.024963
2016-04-03    0.024566
2016-04-04    0.023191
2016-04-05    0.131664
2016-04-06    0.234927
2016-04-07    0.140761
Name: last_seen, dtype: float64

No irregular patterns in distribution. There is no N/A entry. Data cleaning is unnecessary.

In [28]:
autos_filtered['registration_year'].describe()

count    37816.000000
mean      2005.790935
std         87.599466
min       1000.000000
25%       2001.000000
50%       2005.000000
75%       2009.000000
max       9999.000000
Name: registration_year, dtype: float64

There are, however, some clear ourliers in registration years. Year of registration should not be before 1885 (the year the first automobile was made) or after 2016 (latest year in *ad_created*).

We will count the number of rows within our range.

In [29]:
autos_filtered['registration_year'][autos_filtered['registration_year'].between(1885,2016)].count()

36443

There will be 36,443 entries left after filtering. We will now calculate the distribution of the remaining values.

In [30]:
registration_year_filtered = autos_filtered['registration_year'][autos_filtered['registration_year'].between(1885,2016)]
registration_year_filtered.value_counts(normalize = True).sort_index()

1927    0.000027
1929    0.000027
1931    0.000027
1934    0.000055
1937    0.000110
          ...   
2012    0.035809
2013    0.021815
2014    0.017809
2015    0.009851
2016    0.016848
Name: registration_year, Length: 77, dtype: float64

There is no apparent outlier in the distribution of registration years after filtering. The majority of the cars ahve been registerd in the period between 1990 - 2016.

In [31]:
autos_filtered = autos_filtered[autos_filtered['registration_year'].between(1885,2016)]

When working on the data of the used cars, it would also be natural to explore the price variations across different car brands.

In [32]:
autos_filtered['brand'].value_counts().head(10)

volkswagen       7680
bmw              4609
mercedes_benz    4104
audi             3592
opel             3208
ford             2110
renault          1331
peugeot          1001
fiat              759
skoda             701
Name: brand, dtype: int64

We will be finding the average prices of the top 10 most common brands.

In [33]:
common_brand = autos_filtered['brand'].value_counts().head(10)

In [34]:
brand_avg_price = {}
for brand in common_brand.index:
    brand_sum = autos_filtered.loc[autos_filtered['brand']==brand, 'price'].sum()
    brand_count = autos_filtered.loc[autos_filtered['brand']==brand, 'price'].count()
    brand_avg_price[brand] = brand_sum/brand_count

In [35]:
print(brand_avg_price)

{'volkswagen': 6764.713671875, 'bmw': 9105.894771100022, 'mercedes_benz': 9399.551169590643, 'audi': 10423.361080178174, 'opel': 4325.7584164588525, 'ford': 5482.854502369668, 'renault': 3699.3944402704733, 'peugeot': 4063.944055944056, 'fiat': 4106.537549407115, 'skoda': 6903.306704707561}


In [36]:
inv_price = sorted([[v,k] for k,v in brand_avg_price.items()],reverse = True)

for v,k in inv_price:
    print('Brand {} has the average price of ${:,.2f} on eBay.'.format(k.capitalize(),v))

Brand Audi has the average price of $10,423.36 on eBay.
Brand Mercedes_benz has the average price of $9,399.55 on eBay.
Brand Bmw has the average price of $9,105.89 on eBay.
Brand Skoda has the average price of $6,903.31 on eBay.
Brand Volkswagen has the average price of $6,764.71 on eBay.
Brand Ford has the average price of $5,482.85 on eBay.
Brand Opel has the average price of $4,325.76 on eBay.
Brand Fiat has the average price of $4,106.54 on eBay.
Brand Peugeot has the average price of $4,063.94 on eBay.
Brand Renault has the average price of $3,699.39 on eBay.


Audi on avereage has the highest sales price within the dataset, followed by Mercedes and BMW respectively. The pricing for Mercedes, Audi, and BMW are in line with the general public opinions on the price of cars.

There is, however, a distinct price gap between the top three brands and the following brands.

Could there be other factors that affect the pricing for the top three brands, perhaps in terms of mileage? This will be investigated below. We will also construct an individual dataframe containing the mean price and the mean mileage.

In [37]:
brand_avg_mileage = {}
for brand in common_brand.index:
    mileage_sum = autos_filtered.loc[autos_filtered['brand']==brand, 'odometer_km'].sum()
    brand_count = autos_filtered.loc[autos_filtered['brand']==brand, 'odometer_km'].count()
    brand_avg_mileage[brand] = mileage_sum/brand_count

In [38]:
bap_series = pd.Series(brand_avg_price)
bam_series = pd.Series(brand_avg_mileage)

In [39]:
comparison_df = pd.DataFrame(bap_series, columns = ['mean_price'])
comparison_df['mean_mileage'] = bam_series

In [40]:
comparison_df.sort_values('mean_price',ascending = False)

mean_price   mean_mileage
audi           10423.361080  127442.928731
mercedes_benz   9399.551170  129870.857700
bmw             9105.894771  131846.387503
skoda           6903.306705  109643.366619
volkswagen      6764.713672  125420.572917
ford            5482.854502  118907.582938
opel            4325.758416  123424.251870
fiat            4106.537549  107134.387352
peugeot         4063.944056  121608.391608
renault         3699.394440  120732.531931

There does not seem to be obvious correlation between the mean price and the mean mileage across different brands - the used cars on average have between 100km abd 132km mileage clocked before they are listed on the eBay marketplace.